path_oidc.go

```diff
@@ -44,6 +58,107 @@ func pathOIDC(b *jwtAuthBackend) []*framework.Path {
 	}
 }
 
+func (b *jwtAuthBackend) pathCallback(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	state := b.verifyState(d.Raw["state"].(string))
+	if state == nil {
+		return logical.ErrorResponse("expired or missing OAuth state"), nil
+	}
+
+	roleName := state.rolename
+	role, err := b.role(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return logical.ErrorResponse("role could not be found"), nil
+	}
+
+	config, err := b.config(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if config == nil {
+		return logical.ErrorResponse("could not load configuration"), nil
+	}
+
+	provider, err := b.getProvider(ctx, config)
+	if err != nil {
+		return nil, errwrap.Wrapf("error getting provider for login operation: {{err}}", err)
+	}
+
+	var oauth2Config = oauth2.Config{
+		ClientID:     config.OIDCClientID,
+		ClientSecret: config.OIDCClientSecret,
+		RedirectURL:  state.redirectURI,
+		Endpoint:     provider.Endpoint(),
+		Scopes:       []string{oidc.ScopeOpenID},
+	}
+
+	oauth2Token, err := oauth2Config.Exchange(ctx, d.Raw["code"].(string))
+	if err != nil {
+		return nil, errwrap.Wrapf("error exchanging oidc code: {{err}}", err)
+	}
+
+	// Extract the ID Token from OAuth2 token.
+	rawToken, ok := oauth2Token.Extra("id_token").(string)
+	if !ok {
+		return nil, errors.New("no id_token found in response")
+	}
+
+	// Parse and verify ID Token payload.
+	allClaims, err := b.verifyToken(ctx, config, role, rawToken)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
+
+	// Attempt to fetch information from the /userinfo endpoint and merge it with
+	// the existing claims data. A failure to fetch additional information from this
+	// endpoint will not invalidate the authorization flow.
+	if userinfo, err := provider.UserInfo(ctx, oauth2.StaticTokenSource(oauth2Token)); err == nil {
+		userinfo.Claims(&allClaims)
+	} else {
+		b.Logger().Info("error reading /userinfo endpoint", "error", err)
```

Comment:
Should this be at the Warn level? I usually put errors I'm swallowing at Warn but I'm not sure if that makes sense here.

Reply:
Yeah, probably so. I just upped it to Warn by default. I'll still use Info if the error is that the provider simply doesn't support the `/userinfo` endpoint, which isn't really an error to call attention to but just a trait of the provider. I thought there'd be a simple way to know whether they support it before calling the function, but I'm not seeing it.